src/lib/y2storage/encryption_processes/pervasive.rb

```diff
@@ -63,19 +60,30 @@ def pre_commit(device)
         # API to allow sharing a new key among several volumes.
         @secure_key ||= generate_secure_key(device)
 
-        @zkey_cryptsetup_output = execute_zkey_cryptsetup(device)
-        return if @zkey_cryptsetup_output.empty?
+        master_key_file = "/etc/zkey/repository/#{@secure_key.name}.skey"
+        sector_size = sector_size_for(device.blk_device)
 
-        device.format_options = luksformat_options_string + " --pbkdf pbkdf2"
+        device.format_options = "--master-key-file #{master_key_file.shellescape} --key-size 1024 "\
+                                "--cipher paes-xts-plain64"
+        device.format_options += " --sector-size #{sector_size}" if sector_size
+        device.format_options += " --pbkdf pbkdf2"
       end
 
       # @see Base#post_commit
       #
-      # Executes the extra commands reported by the former call to
-      # "zkey cryptsetup".
+      # Adds the device to the secure key if needed and executes the
+      # extra commands reported by a call to "zkey cryptsetup".
       #
       # @param device [Encryption] encryption that has just been created in the system
       def post_commit(device)
+        if !@secure_key.for_device?(device)
+          secure_key_volume = @secure_key.add_device(device)
+
+          Yast::Execute.locally(ZKEY, "change", "--name", secure_key.name, "--volumes",
+            "+#{secure_key_volume.plain_name.shellescape}:#{secure_key_volume.dm_name.shellescape}")
+        end
```

Comment:
This looks like logic that should live in the SecureKey class. That's the class that contains all the logic to register the key in the system (by calling "zkey generate") or to copy the keys to the final system at the end of the installation. Having "zkey" commands here to modify in the system the key generated by the `SecureKey` object without telling that to that object looks quite wrong.

Reply:
But the SecureKey class has no concept of whether the device is added to the key only in memory or on-disk. Prior to the call to add the volume on-disk is a call to the class to add it in-memory. So that looks fine to me.

Reply:
So maybe we should make the `SecureKey` class smarter or change its API a bit. But generating the system key inside the object, then modifying it from outside and finally using the object again to copy the system key to the final system cannot be fine, IMO.

Moreoever, in the initial implementation, the `SecureKeyVolume` object that is stored inside the `SecureKey` object encapsulated the logic about how the "zkey" command expresses the mapping of devices ("plain_name:dm_name"). Now we are bypassing all that, directly constructing the "zkey" command out of everything. That's why the coverage decreased, because the class that was created to contain the logic about how "zkey" stores the devices is not being asked for that information any longer.

In general, this change is quite disruptive with the original distribution of responsibilities among `Pervasive`, `SecureKey` and `SecureKeyVolume`.

Reply:
I have just seen that the second part of my comment has been addressed by a subsequent commit.